File: crypto_agent/exchange/manager.py
from .base import BaseExchange


class ExchangeManager:
    def __init__(self):
        self._exchanges: dict[str, BaseExchange] = {}
        self._active_id: str | None = None

    def register(self, exchange_id: str, exchange: BaseExchange):
        self._exchanges[exchange_id] = exchange
        if self._active_id is None:
            self._active_id = exchange_id

    @property
    def active(self) -> BaseExchange:
        if self._active_id is None or self._active_id not in self._exchanges:
            raise RuntimeError("No active exchange. Register one first.")
        return self._exchanges[self._active_id]

    @property
    def active_id(self) -> str:
        return self._active_id or ""

    def set_active(self, exchange_id: str):
        if exchange_id not in self._exchanges:
            raise KeyError(f"Exchange '{exchange_id}' not registered. Available: {list(self._exchanges.keys())}")
        self._active_id = exchange_id

    def get(self, exchange_id: str) -> BaseExchange:
        if exchange_id not in self._exchanges:
            raise KeyError(f"Exchange '{exchange_id}' not registered.")
        return self._exchanges[exchange_id]

    def list(self) -> list[str]:
        return list(self._exchanges.keys())

    async def close_all(self):
        for ex in self._exchanges.values():
            await ex.close()

File: src/exchange/manager.ts
import type { BaseExchange } from "./base.js";

export class ExchangeManager {
  private exchanges: Record<string, BaseExchange> = {};
  private _activeId: string | null = null;

  register(exchangeId: string, exchange: BaseExchange): void {
    this.exchanges[exchangeId] = exchange;
    if (this._activeId === null) {
      this._activeId = exchangeId;
    }
  }

  get active(): BaseExchange {
    if (!this._activeId || !(this._activeId in this.exchanges)) {
      throw new Error("No active exchange. Register one first.");
    }
    return this.exchanges[this._activeId];
  }

  get activeId(): string {
    return this._activeId ?? "";
  }

  setActive(exchangeId: string): void {
    if (!(exchangeId in this.exchanges)) {
      throw new Error(
        `Exchange '${exchangeId}' not registered. Available: ${Object.keys(this.exchanges).join(", ")}`,
      );
    }
    this._activeId = exchangeId;
  }

  get(exchangeId: string): BaseExchange {
    if (!(exchangeId in this.exchanges)) {
      throw new Error(`Exchange '${exchangeId}' not registered.`);
    }
    return this.exchanges[exchangeId];
  }

  list(): string[] {
    return Object.keys(this.exchanges);
  }

  async closeAll(): Promise<void> {
    for (const ex of Object.values(this.exchanges)) {
      await ex.close();
    }
  }
}
